core/iwasm/interpreter/wasm_runtime.c

```diff
@@ -426,12 +509,29 @@ memories_instantiate(const WASMModule *module, WASMModuleInstance *module_inst,
         else
 #endif
         {
-            if (!(memories[mem_index] = memory_instantiate(
-                      module_inst, parent, memory, mem_index,
-                      num_bytes_per_page, init_page_count, max_page_count,
-                      actual_heap_size, flags, error_buf, error_buf_size))) {
-                memories_deinstantiate(module_inst, memories, memory_count);
-                return NULL;
+#if WASM_ENABLE_MULTI_MEMORY != 0
+            /* Current assumption is if there doesn't exist a non-import memory,
+             * the first non-import memory would be default memory */
```

Comment:
i failed to understand this comment.

Reply:
So I am assuming that if there are multiple wasm linear memories, there will be one "main(default)" memory. Some data members like heap_base, data_end, and host-managed heap are only available in default memory. Other memories don't have such data members, that's why I have an extra function that contains some duplicate code to handle different cases.

Reply:
"the default memory" makes me think the memory with index==0, regardless of import/non-import.
and i guess it's where we should put the app heap. (well, at least until we decide how to fix https://github.com/bytecodealliance/wasm-micro-runtime/issues/1353)

> that's why I have an extra function that contains some duplicate code to handle different cases.

i prefer to have `bool try_to_insert_app_heap` or something along the line over having two very similar functions like these.

Reply:
so how about using the term "main memory"

Reply:
why do you need such a concept?
the app heap should be in the default memory, shouldn't it?

Reply:
Hi, @yamt I think you have a valid point, the app heap in the default memory(index 0) is good enough. I overthink and complex the situation here. Let me list all the situations here for reference:
1. If memory 0 is non-import memory, instantiate the memory as usual. 
2. If memory 0 is import memory, 
    - either it's imported from wasm module(app heap initialized when that wasm submodule memory is instantiated) 
    - or it's imported from the builtin module(like "env" or "spectest" that runtime provides), in that case, it will be instantiated as usual

So it's enough to use the default memory to insert the app heap. What do you think?

Reply:
> So it's enough to use the default memory to insert the app heap. What do you think?

it's more than "enough". it's actually how it should be, i guess.

the app heap in non-default memory is almost useless because it can't be addressed by wasm modules written in ordinal languages like C.

Reply:
I agree

Reply:
@yamt so is there any other comment from you? I guess we can merge it to dev/multi_memory for more testing and merge to main sometime in the future?

Reply:
> @yamt so is there any other comment from you?

no other comments.
(i have no time to read the latest version of this PR today.)

> I guess we can merge it to dev/multi_memory for more testing and merge to main sometime in the future?

i guess it's fine.

Reply:
OK, thanks.